### PR TITLE
Remove subdir in PackageRecord pkey

### LINKED ISSUE #9116: conda 4.6.x and conda 4.7.x not able to resolve identical package in different subdirectories
Current Behavior

The following happens:
```
[root@596ad8f0842a /]# conda create -n conda-env --dry-run -- test-package
Collecting package metadata (current_repodata.json): done
Solving environment: failed with current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Package pandas conflicts for:
test-package -> pandas
Package typing conflicts for:
test-package -> typing
Package libffi conflicts for:
test-package -> libffi[version='>=3.2.1,<4.0a0']
Package pyyaml conflicts for:
test-package -> pyyaml
Package six conflicts for:
test-package -> six[version='>=1.12.0']
Package py4j conflicts for:
test-package -> py4j
Package icu conflicts for:
test-package -> icu
Package wrapt conflicts for:
test-package -> wrapt
Package pytz conflicts for:
test-package -> pytz
Package python conflicts for:
test-package -> python[version='>=2.7.14']
Package pyarrow conflicts for:
test-package -> pyarrow
Package matplotlib conflicts for:
test-package -> matplotlib[version='>=2.0.2,<2.3.0']
```
Also observed on conda 4.6.14

### Steps to Reproduce

1. Create a conda repository `test-channel` with the following set up
2. In /linux-64/repodata.json
```
{
  "info": {
    "subdir": "linux-64"
  },
  "packages": {
    "test-package-4.2.1-py_0.tar.bz2": {
      "build": "py_0",
      "build_number": 0,
      "depends": [
        "icu",
        "libffi >=3.2.1,<4.0a0",
        "matplotlib >=2.0.2,<2.3.0",
        "pandas",
        "py4j",
        "pyarrow",
        "python >=2.7.14",
        "pytz",
        "pyyaml",
        "six >=1.12.0",
        "typing",
        "wrapt"
      ],
      "md5": "1512a7e9a56ecb8523475c891bfa6657",
      "name": "test-package",
      "noarch": "python",
      "sha256": "3c02f947500c13ceb8e8aa38db2af43e9017c4948905d0cbd1610185318fe7d9",
      "size": 30098,
      "subdir": "noarch",
      "timestamp": 1565634321434,
      "version": "4.2.1"
    }
  },
  "removed": [],
  "repodata_version": 1
}
```
3. In /noarch/repodata.json
```
{
  "info": {
    "subdir": "noarch"
  },
  "packages": {
    "test-package-4.2.1-py_0.tar.bz2": {
      "build": "py_0",
      "build_number": 0,
      "depends": [
        "icu",
        "libffi >=3.2.1,<4.0a0",
        "matplotlib >=2.0.2,<2.3.0",
        "pandas",
        "py4j",
        "pyarrow",
        "python >=2.7.14",
        "pytz",
        "pyyaml",
        "six >=1.12.0",
        "typing",
        "wrapt"
      ],
      "md5": "1512a7e9a56ecb8523475c891bfa6657",
      "name": "test-package",
      "noarch": "python",
      "sha256": "3c02f947500c13ceb8e8aa38db2af43e9017c4948905d0cbd1610185318fe7d9",
      "size": 30098,
      "subdir": "noarch",
      "timestamp": 1565634321434,
      "version": "4.2.1"
    }
  },
  "removed": [],
  "repodata_version": 1
}
```
3. Set up conda to only point to the test-channel, note I see this error on `--dry-run` so I did not populate the test-channel with the actual packages.
4. Run `conda create -n conda-env --dry-run -- test-package`

## Expected Behavior

Previously, in conda 4.5.x, there was no issue with resolving packages like this.

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
[root@596ad8f0842a /]# conda info

     active environment : None
       user config file : /root/.condarc
 populated config files : /root/.condarc
                          /condarc
          conda version : 4.7.10
    conda-build version : not installed
         python version : 3.7.3.final.0
       virtual packages : 
       base environment : /miniconda  (writable)
           channel URLs : file://test-channel/asset/conda/linux-64
                          file://test-channel/asset/conda/noarch
          package cache : /miniconda/pkgs
                          /root/.conda/pkgs
       envs directories : /miniconda/envs
                          /root/.conda/envs
               platform : linux-64
             user-agent : conda/4.7.10 requests/2.22.0 CPython/3.7.3 Linux/4.9.184-linuxkit centos/6.10 glibc/2.12
                UID:GID : 0:0
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
[root@596ad8f0842a /]# conda config --show-sources
==> /root/.condarc <==
channel_priority: disabled
channels:
  - file://test-channel/asset/conda/
repodata_fns:
  - repodata.json
use_only_tar_bz2: False

==> /condarc <==
channels: []
default_channels: []
```
</p></details>
Note I attempted to remove current_repodata.json from being considered, but this is 4.7.10 before the config value is apparently respected.

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
[root@596ad8f0842a /]# conda list --show-channel-urls
# packages in environment at /miniconda:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    https://repo.anaconda.com/pkgs/main
asn1crypto                0.24.0                   py37_0    https://repo.anaconda.com/pkgs/main
bzip2                     1.0.8                h7b6447c_0    https://repo.anaconda.com/pkgs/main
ca-certificates           2019.5.15                     0    https://repo.anaconda.com/pkgs/main
certifi                   2019.6.16                py37_0    https://repo.anaconda.com/pkgs/main
cffi                      1.12.3           py37h2e261b9_0    https://repo.anaconda.com/pkgs/main
chardet                   3.0.4                    py37_1    https://repo.anaconda.com/pkgs/main
conda                     4.7.10                   py37_0    https://repo.anaconda.com/pkgs/main
conda-package-handling    1.3.11                   py37_0    https://repo.anaconda.com/pkgs/main
cryptography              2.7              py37h1ba5d50_0    https://repo.anaconda.com/pkgs/main
idna                      2.8                      py37_0    https://repo.anaconda.com/pkgs/main
libarchive                3.3.3                h5d8350f_5    https://repo.anaconda.com/pkgs/main
libedit                   3.1.20181209         hc058e9b_0    https://repo.anaconda.com/pkgs/main
libffi                    3.2.1                hd88cf55_4    https://repo.anaconda.com/pkgs/main
libgcc-ng                 9.1.0                hdf63c60_0    https://repo.anaconda.com/pkgs/main
libstdcxx-ng              9.1.0                hdf63c60_0    https://repo.anaconda.com/pkgs/main
libxml2                   2.9.9                hea5a465_1    https://repo.anaconda.com/pkgs/main
lz4-c                     1.8.1.2              h14c3975_0    https://repo.anaconda.com/pkgs/main
lzo                       2.10                 h49e0be7_2    https://repo.anaconda.com/pkgs/main
ncurses                   6.1                  he6710b0_1    https://repo.anaconda.com/pkgs/main
openssl                   1.1.1c               h7b6447c_1    https://repo.anaconda.com/pkgs/main
pip                       19.1.1                   py37_0    https://repo.anaconda.com/pkgs/main
pycosat                   0.6.3            py37h14c3975_0    https://repo.anaconda.com/pkgs/main
pycparser                 2.19                     py37_0    https://repo.anaconda.com/pkgs/main
pyopenssl                 19.0.0                   py37_0    https://repo.anaconda.com/pkgs/main
pysocks                   1.7.0                    py37_0    https://repo.anaconda.com/pkgs/main
python                    3.7.3                h0371630_0    https://repo.anaconda.com/pkgs/main
python-libarchive-c       2.8                     py37_11    https://repo.anaconda.com/pkgs/main
readline                  7.0                  h7b6447c_5    https://repo.anaconda.com/pkgs/main
requests                  2.22.0                   py37_0    https://repo.anaconda.com/pkgs/main
ruamel_yaml               0.15.46          py37h14c3975_0    https://repo.anaconda.com/pkgs/main
setuptools                41.0.1                   py37_0    https://repo.anaconda.com/pkgs/main
six                       1.12.0                   py37_0    https://repo.anaconda.com/pkgs/main
sqlite                    3.29.0               h7b6447c_0    https://repo.anaconda.com/pkgs/main
tk                        8.6.8                hbc83047_0    https://repo.anaconda.com/pkgs/main
tqdm                      4.32.1                     py_0    https://repo.anaconda.com/pkgs/main
urllib3                   1.24.2                   py37_0    https://repo.anaconda.com/pkgs/main
wheel                     0.33.4                   py37_0    https://repo.anaconda.com/pkgs/main
xz                        5.2.4                h14c3975_4    https://repo.anaconda.com/pkgs/main
yaml                      0.1.7                had09818_2    https://repo.anaconda.com/pkgs/main
zlib                      1.2.11               h7b6447c_3    https://repo.anaconda.com/pkgs/main
zstd                      1.3.7                h0b5b093_0    https://repo.anaconda.com/pkgs/main
```
</p></details>

### PR DESCRIPTION
**Pull request by jeremyjliu**
Fixes #9116.

Since conda 4.6, the switch to using `PackageRecord` over `Dist` in `resolve.py` means otherwise identical packages in different subdirs are considered distinct, which breaks the solver when given this input as reduced_index.* Previously in the `Dist` object, these would not be considered distinct and one would be arbitrarily chosen over the other for input into the solver, so the bug is considered a regression.

This PR removes the `subdir` parameter in `PackageRecord._pkey` which is utilized to compute equality, allowing the same behavior as before. This is especially important in here: https://github.com/conda/conda/blob/master/conda/resolve.py#L753 when conda prepares the index for the solve.

*Note: I am not sure why the solver is unable to solve this case when it feels like it should. I opted to file this fix first for discussion before investigating further.

### CODE REVIEW

**Comment by msarahan** [comment]
Thanks @jeremyjliu!  I think I agree with this change, but I have one concern: determinism.  We have worked recently to ensure that conda has deterministic behavior when a package is totally identical in noarch and somearch-XY.  There's a solver step that does that.  I think your way here is probably more efficient, but I want to understand it better.  Right now, I think you leave it more or less up to chance.  This is especially true because the repodata.json files are read in parallel, though I think with the latest map approach, the order should still be predictable.

TL;DR: I think we need to verify that subdir_data.py has deterministic loading behavior, and that the somearch-XY is always preferred over the noarch package with otherwise similar info.  That means that somearch-XY either needs to be processed first, or that it needs to be processed last.  My memory is a bit shaky here, but I think it is the former: the first entry into the index is the one that "wins" and there is no clobbering by later entries.

**Comment by msarahan** [comment]
@cla-bot recheck

**Comment by msarahan** [comment]
@cla-bot check

**Comment by msarahan** [comment]
retargeted to 4.7.x for 4.7.13 bugfix release

**Comment by kalefranz** [comment]
I’d raise caution against this change, as I’m not sure it’s the right way to solve the problem being observed. The packages between an arch-specific subdir and noarch are only equivalent if their badges and the contents of their repodata are exactly the same. That’s pretty much never the case. Since in 4.3.27 we implemented channel priority even within multi-channels like `defaults`, it seems to me the correct solution is to prioritize not just channels, but actually doing an outer join on `context.channels` + `context.subdirs`, and then prioritizing the packages within in that order. (And preferably via strict priority.)

**Comment by jeremyjliu** [comment]
Hi @msarahan and @kalefranz, thanks for the feedback! Hm, I agree with you on both points, in that I think it's better to surface each `subdir` variant of a package to the solver and solve in a deterministic way. I investigated a bit further and opened https://github.com/conda/conda/pull/9261 as an alternative to this PR, to allow the solver to handle otherwise identical packages with different `subdir`.

**Comment by jeremyjliu** [comment]
Closing in favor of https://github.com/conda/conda/pull/9261
